pkg/ioutil/ioutil.go

```diff
@@ -205,47 +205,55 @@ func CopyAligned(w *os.File, r io.Reader, alignedBuf []byte) (int64, error) {
 	}
 
 	var written int64
-	var err error
 	for {
-		nr, er := r.Read(alignedBuf)
-		if nr == len(alignedBuf) {
-			// Buffer read is aligned with input buffer, proceed to write.
-			nw, ew := w.Write(alignedBuf)
+		buf := alignedBuf
+		if totalSize != -1 {
+			remaining := totalSize - written
+			if remaining < int64(len(buf)) {
+				buf = buf[:remaining]
+			}
+		}
+		nr, err := io.ReadFull(r, buf)
```

Comment:
Is this the place where we read 4MB from the network? If the input rate is slow, waiting to fill a 4MB buffer is likely to slow down performance even further.

Reply:
> Is this the place where we read 4MB from the network?

yes

> If the input rate is slow, waiting to fill a 4MB buffer is likely to slow down performance even further.

No, because instead of the data going to the disk, the data will go into this buffer. This will have a better effect on the latency. Once the buffer is full, it will write to the disk in a bigger block size of 4MB which will help with seq throughput.

Reply:
@donatello slowness is only for a situation where we see premature EOF, it seems like Pipe() becomes slow at EOF.

This PR is a fix when we know the exact size of the reader, but @krishnasrinivas needs to do some numbers before we can make that assumption.